core/scale/tie.hpp

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_UTILS_SCALE_TUPLE_HPP
+#define KAGOME_UTILS_SCALE_TUPLE_HPP
+
+#include <tuple>
+
+#define SCALE_TIE(N) static constexpr size_t scale_tie = N
+#define SCALE_TIE_EQ(T)                                       \
+  bool operator==(const T &r) const {                         \
+    static_assert(std::is_same_v<const T *, decltype(this)>); \
+    return ::scale::as_tie(*this) == ::scale::as_tie(r);      \
+  }                                                           \
+  bool operator!=(const T &r) const {                         \
+    static_assert(std::is_same_v<const T *, decltype(this)>); \
+    return !operator==(r);                                    \
+  }
+
+namespace scale {
+  class ScaleEncoderStream;
+  class ScaleDecoderStream;
+
+  // generated by housekeeping/scale_tie.py
+  template <typename T, size_t N = std::remove_reference_t<T>::scale_tie>
+  auto as_tie(T &&v) {
```

Comment:
It's not rvalue, it'universal reference.
It will match both "T &" and "const T &" (avoids method duplication)

Reply:
Yep, I know. And it is meaningless here. Also your code results to ```std::tuple<T1&, T2&,...>```. Better to have ```std::tuple<T1 const &, T2 const &,...>```. To do so, you need to declare structure bindings as ```const```.

Reply:
Moreover...it seems it will be better to make ```as_tie(T &)```. Because it avoids the posibility of such a call
```as_tie(SomeStruct{...})```. Because after function returns, the scope will be closed and SomeStruct will be destroyed. And references in tuple will point to dead memory.

Reply:
Yep, I know. But it is meaningless here. Your function returns ```std::tuple<T1&, T2&>``` instead of ```std::tuple<T1 const&, T2 const&>```. Moreover it allows to make a call like ```as_tie(SomeStruct{})```. Which causes to have a tuple with references to dead objects! So we should make ```as_tie(T &v)``` to avoid such kind of calls.

Reply:
Sorry, i don't see that it will be used in decoding process...So yep, ```const``` structured binding should be skipped. But ```as_tie(T &v)``` still should be.

Reply:
Let's do smth like this:
```c++
template <typename T, typename F, size_t N = std::remove_reference_t<T>::scale_tie>
auto as_tie(T &&v, F &&func) {
  if constexpr (N == 1) {
    auto &[v0] = v;
    return std::forward<F>(func)(std::tie(v0));
  }
  ...
  else {
    // generate code for bigger tuples
    (void)0;
  }
}

template <typename T, size_t = T::scale_tie>
ScaleEncoderStream &operator<<(ScaleEncoderStream &s, const T &v) {
  return as_tie(v, [&](auto &&t) {
    return s << t;
  });
}

template <typename T, size_t = T::scale_tie>
ScaleDecoderStream &operator>>(ScaleDecoderStream &s, T &v) {
  return as_tie(v, [&](auto &&t) {
    decode_tie<0>(s, t);
    return s;
  });
}